### ISSUE TITLE
xbuild in FSharpBuildProject ignores g:fsharp_build_path

### ISSUE THREAD

**Issue by yever** [fsharp/vim-fsharp#23]
They way it's coded now, it's impossible to build a project unless "xbuild" is in the path. 
The variable ``g:fsharp_build_path`` is ignored in: https://github.com/fsharp/vim-fsharp/blob/master/autoload/fsharpbinding/python.vim#L70.

It would be nice to enable the use of MSBuild in Windows. It's pretty easy to get its path with:
```dos
reg.exe query "HKLM\SOFTWARE\Microsoft\MSBuild\ToolsVersions\4.0" /v MSBuildToolsPath
```
(taken from [this answer in stackoverflow](http://stackoverflow.com/questions/328017/path-to-msbuild/15446618#15446618)). Maybe even better like so:
```dos
reg.exe query "HKLM\SOFTWARE\Microsoft\MSBuild\ToolsVersions" /v MSBuildToolsPath /s
```
in order to get all available versions.

**Comment by kjnilsson** [comment]
Hi,

Thanks for the report - seems like we got some legacy code left - the official way for building the project now is to use :make which should heed the fsharp_xbuild_path setting. We need to change the code so that the old commands use the same code. Until then you should be able to use :make to build the currently scoped project.

Cheers
Karl

**Comment by kjnilsson** [comment]
I'd also welcome pull requests that implemented registry scanning on windows here: https://github.com/fsharp/vim-fsharp/blob/master/ftplugin/fsharp.vim#L16

:)

**Comment by yever** [comment]
Hi @kjnilsson :)

Thank you for you quick response. I must say that I am really amazed by this plugin and the power it gives you with IntelliSense, type checking, etc. (Though I still haven't figured out how to make it find the project references and get the dependencies right.)

To the point, it indeed makes more sense to use ``:make``, but I guess it wouldn't do any harm to update ``FSharpBuildProject`` to work the same way. I'll go ahead and make a pull request addressing this and the registry scanning for Windows.

I'm wondering about still referring to the 'build' command as 'xbuild' in the name of the global variable, even though that's the Mono one. Strictly speaking, Mono can also be installed on Windows...

**Comment by yever** [comment]
Come to think of it, maybe it would be easier to just let ``FSharpBuildProject`` be an alias to ``:make`` (if backward compatibility is required). I'm having a hard time escaping ``g:fsharp_xbuild_path`` correctly to make it work the same way in both cases.

**Comment by kjnilsson** [comment]
I agree the name isn't great. We could change it but we would still have to check the old value to at least see if it is set and print a warning.

Can you elaborate on your problems finding the project references and dependencies?

**Comment by yever** [comment]
I managed to figure out the references/dependencies issue. Vim should be started in the directory where the solution/project file is...

After implementing the registry scanning, I'm thinking that this might not be the right solution after all. For one, this won't work in Cygwin (using its native vim/gvim) since Cygwin's python doesn't include the module ``_winreg``/``winreg``. This can be overcome, but it's getting quite hacky.

This project seems very unix-centric as it is now. Everything is executed with ``mono`` (see the ``execute '!mono ' ...`` lines in [python.vim](https://github.com/fsharp/vim-fsharp/blob/master/autoload/fsharpbinding/python.vim)). This seems in line with @kongo2002's [recommendation](https://github.com/kongo2002/fsharp-vim/blob/master/README.mkd) to use Visual Studio with VsVim under Windows.

I personally find it useful to still have the choice to use vim under Windows. In the [readme file](https://github.com/fsharp/vim-fsharp) of this repository there are installation instructions for Windows, so I guess I'm not completely off the track in trying to get it to work.

From a different perspective, MSBuild now has [cross-platform support](https://github.com/Microsoft/msbuild/issues/40). I haven't tested it yet myself, but if it does work well, it might make sense to bypass ``xbuild`` altogether and add NuGets for [MSBuild](https://www.nuget.org/packages/MSBuild/) and the [F# compiler tools (including FSI)](https://www.nuget.org/packages/FSharp.Compiler.Tools/). This would make the plugin heavier but at the same time platform-independent. It would be more on the side of turning vim into an IDE that includes the compiler in it.

**Comment by kjnilsson** [comment]
You shouldn't have to start vim in any particular location. if you open an .fs file vim will check if there is an fsproj file in the same directory as the file and parse that for references.

I still think registry scanning would be a useful addition. For cygwin use users can manually specify the msbuild path for now.

Yes some of the command have a hard-coded 'mono ' prefix - it should be fairly straight-forward to change this to do the right thing on windows. I'll look into it.

Cheers
Karl

**Comment by yever** [comment]
Hello Karl,

It seems like this conversation is growing beyond its original scope. I'll open a new issue regarding references. I'm still getting some weird behavior.

Regarding the registry, wouldn't it be better to rather address this issue in the install script? It could make sense to check if there's an already installed build system (MSBuild or xbuild) and use it; otherwise, MSBuild can be installed with a NuGet. A FAKE script will have the advantage of having access to [``EnvironmentHelper.isMono``](https://github.com/fsharp/FAKE/blob/master/src/app/FakeLib/EnvironmentHelper.fs#L128-128). The less mainstream options of having mono on Windows or wishing to install MSBuild on mono can be left out for the time being, or be enabled with an optional parameter. The registry scanning itself can be done in the FAKE script, using the .NET functions meant for it (I found [this snippet](http://www.fssnip.net/1m), I guess it could be done in a more F#-y way).

The hard-coded ``mono`` prefix can probably also go into a new issue regarding cross-platform support.

Best,
Ya'ar

**Comment by kjnilsson** [comment]
Doing it in the fake script sounds like good idea. How would you configure the vim plugin to use what the build script found

**Comment by yever** [comment]
I'm trying to figure it out... 

What do you think about using the [StringHelper](http://fsharp.github.io/FAKE/apidocs/fake-stringhelper.html)? It has a ``ReplaceInFile`` function that replaces the contents of a file and an operator ``(>=>)`` for RegEx manipulations. It could be used for templating.

Would it be OK to change ``g:fsharp_xbuild_path`` to ``g:fsharp_build_system_path``? Should there be backward-compatibility here?

I would still go for
```
if !exists('g:fsharp_build_system_path')
    let g:fsharp_build_system_path = {BuildSystem}
```
with ``{BuildSystem}`` used as a template for what the FAKE script found -- to allow the user to overwrite the path in ``.vimrc``.

PS: I just found out that FAKE has a [RegistryHelper](http://fsharp.github.io/FAKE/apidocs/fake-registryhelper.html). This could come in handy.

**Comment by yever** [comment]
Hi @kjnilsson. I'm sorry, I was a bit under the weather and things took longer than expected.
Anyway, here is a new [pull request](https://github.com/fsharp/vim-fsharp/pull/26). I have tested it and it seems to work fine.

Next on the agenda would be:
- to change ``g:fsharp_xbuild_path`` to a name not involving "xbuild"
- to add something similar for finding the ``fsi`` executable (I'm thinking about exploiting fsautocomplete)
- to update the version of fsautocomplete (I tried it and it doesn't work...)
